tests/bgp: Add mgmtd set-src regression coverage

Bojun-Feng opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/sonic-net/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background context?
- List any dependencies that are required for this change.
-->

Adds regression coverage for the mgmtd FRR replay gap that previously dropped the default-route `set src` entry when configs grew large. 

The new file `tests/bgp/test_frr_set_src_mgmtd.py` verifies that both kernel default routes and FRR route-maps continue pointing to Loopback0 after a standard mgmtd reload and after a reload preceded by injecting 512 static routes via `sonic-cfggen`. Both tests run on T0/T1 hardware and are gated on `frr_mgmt_framework_config`.

Fix #21342
Fix https://github.com/sonic-net/sonic-buildimage/issues/24694

Testing regression for https://github.com/sonic-net/sonic-buildimage/issues/21931
Original issue https://github.com/FRRouting/frr/issues/18541 is fixed in upstream by https://github.com/FRRouting/frr/pull/18601

### Type of change

<!--
- Fill x for your type of change.
- e.g.
- [x] Bug fix
-->

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] New Test case
    - [ ] Skipped for non-supported platforms
- [ ] Test case improvement


### Back port request
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [ ] 202405
- [ ] 202411
- [ ] 202505

### Approach
#### What is the motivation for this PR?
Prevent regressions where mgmtd’s FRR replay loses the default-route source IP (Loopback0), especially when replaying long configs.

#### How did you do it?
- Added helpers to read Loopback0 per ASIC and validate both kernel and FRR state.
- Implemented two mgmtd tests: one plain reload, one reload after injecting 512 static routes (with checkpoint/rollback cleanup).
- Required `frr_mgmt_framework_config` so the tests only run where mgmtd is enabled.

#### How did you verify/test it?
- `python3 -m compileall tests/bgp/test_frr_set_src_mgmtd.py`
- `pytest tests/bgp/test_frr_set_src_mgmtd.py -k mgmtd_preserves_default_route_set_src_with_large_config`

#### Any platform specific information?
Runs on hardware T0/T1 (skipped elsewhere); no platform-specific logic.

#### Supported testbed topology if it's a new test case?
`t0`, `t1`

### Documentation
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->
I don't think updating documentation is necessary for this one? Please correct me if I'm wrong.

Bojun-Feng committed:
tests/bgp: Add mgmtd set-src regression coverage

* Add helpers to validate Loopback0 source IP on all ASICs (kernel routes + FRR route-maps)
* Add mgmtd reload test plus a long-config variant that injects 512 static routes via sonic-cfggen
* Gate both tests on frr_mgmt_framework_config and ensure checkpoint/rollback cleanup after bloat

Signed-off-by: Bojun Feng <bojundf@gmail.com>

mssonicbld commented:
/azp run

Gfrom2016 reviewed (dismissed):
LGTM. Clean regression test — good coverage of both the simple reload case and the large-config scenario that triggered the original mgmtd replay bug. Multi-ASIC support and checkpoint/rollback cleanup look solid.

StormLiangMS reviewed (dismissed):
**LGTM** — Clean, well-structured regression test for the mgmtd FRR replay bug.

**What's good:**
- Two-tier coverage: simple reload + large-config (512 static routes) stress test
- Proper `frr_mgmt_framework_config` gating
- Multi-ASIC support with frontend ASIC iteration
- Clean checkpoint/rollback cleanup in `try/finally`
- Iteration-level support for varying CI depth

**Minor nits (non-blocking):**
1. `_get_remote_ip()` — the `except StopIteration` is dead code (`for` loop never raises it); harmless but could simplify
2. `test_...large_config` — `checkpoint_created` flag could be collapsed by moving `create_checkpoint` inside the `try` block
3. Missing trailing newline at EOF

None of these block merge. Approve.

StormLiangMS commented:
@deepak-singhal0408 would you help to review and approve?

Bojun-Feng committed:
[test]: Address review nits in mgmtd set-src regression test

* Remove dead except StopIteration in _get_remote_ip (for loop never raises it)
* Move create_checkpoint inside try block, eliminating checkpoint_created flag

Signed-off-by: Bojun Feng <bojundf@gmail.com>

mssonicbld commented:
/azp run

Bojun-Feng commented:
/azpw retry

mssonicbld commented:
Retrying failed(or canceled) jobs...

mssonicbld commented:
Retrying failed(or canceled) stages in build 1099517:


✅Stage **Test**:
- Job impacted-area-kvmtest-t0-sonic by Elastictest: retried.

Bojun-Feng commented:
Hi @deepak-singhal0408 — would you be able to review this when you get a chance? Just finished addressing all of the nits. CI is green. Let me know if there is anything else needed from me, thanks!

deepak-singhal0408 commented on tests/bgp/test_frr_set_src_mgmtd.py:
```diff
@@ -0,0 +1,244 @@
+    """
+    duthost = duthosts[enum_rand_one_per_hwsku_frontend_hostname]
+
+    pytest_require(
```
`frr_mgmt_framework_config` controls SONiC's native mgmtd API integration (gNMI/REST through mgmtd). It does **not** control whether `mgmtd` runs — mgmtd is mandatory in all FRR 10.x deployments regardless of this flag. `vtysh` routes all YANG-modeled config through mgmtd internally.

Evidence from [sonic-buildimage#21931](https://github.com/sonic-net/sonic-buildimage/issues/21931): dgsudharsan confirmed reproduction in **unified mode** (without `frr_mgmt_framework_config`). StormLiangMS's GDB trace was also on a standard-mode device. Any SONiC deployment on FRR 10.x (202411+) has mgmtd running and is affected.

This skip condition means the test will be **skipped on most production-relevant devices** that don't have this flag enabled but still run mgmtd.

Suggested fix — gate on mgmtd process presence instead:
```python
mgmtd_running = duthost.shell(
    "docker exec bgp pgrep -x mgmtd", module_ignore_errors=True
)['rc'] == 0
pytest_require(mgmtd_running, "Test requires mgmtd (FRR 10.x+)")
```
This same issue applies to the second test function at line 220.

Bojun-Feng replied:
Feedback 1, resolved by https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/8b8a44ed3663529d1c2feccaa6deb1bd8125d5aa

Now we are checking `pgrep -x mgmtd` in bgp containers as suggested.

deepak-singhal0408 reviewed (commented):
Review based on deep analysis of the underlying bug ([sonic-buildimage#21931](https://github.com/sonic-net/sonic-buildimage/issues/21931), [FRRouting/frr#18541](https://github.com/FRRouting/frr/issues/18541)) and upstream fix ([FRRouting/frr#18601](https://github.com/FRRouting/frr/pull/18601)).

deepak-singhal0408 commented on tests/bgp/test_frr_set_src_mgmtd.py:
```diff
@@ -0,0 +1,244 @@
+
+pytestmark = [
+    pytest.mark.disable_loganalyzer,
+    pytest.mark.topology('t0', 't1')
```
The bug is topology-agnostic — it's an FRR-internal race between mgmtd sessions that applies to every device running FRR 10.x with bgpcfgd, including T2 chassis. The code already has multi-ASIC support via `_get_asic_hosts()` (iterates all frontend ASICs), but the topology marker prevents it from ever running on T2.

```python
pytest.mark.topology('t0', 't1', 't2')
```

Bojun-Feng replied:
Feedback 2, resolved by https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/8b8a44ed3663529d1c2feccaa6deb1bd8125d5aa

Added `t2` topology as suggested.

Bojun-Feng replied:
One nuance: On T2 chassis, the default route is only advertised from T3 to a single linecard. I used the `dut_with_default_route` fixture to ensure we target a linecard with the route, allowing us to verify that `set_src` is applied to routes rather than just appearing in the config. 

The trade-off is that we won't catch failures exclusive to linecards without a default route. However, since the bug is FRR-internal, this should be sufficient.

Please let me know if I've overlooked any concerns regarding this approach.

We switched to `dut_with_default_route` in commit https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/4c771a8d4cf0ba4cfa273fac72a6d99892a2b5a0
For more detail see comment in fixture definition:
https://github.com/sonic-net/sonic-mgmt/blob/9357b25ef559ffef8abd831b79ade3070528af17/tests/bgp/conftest.py#L862-L881

deepak-singhal0408 commented on tests/bgp/test_frr_set_src_mgmtd.py:
```diff
@@ -0,0 +1,244 @@
+        "Test requires FRR mgmt-framework (mgmtd) mode to be enabled"
+    )
+
+    normalized_level = get_function_completeness_level if get_function_completeness_level else 'debug'
```
The production trigger ([sonic-buildimage#21931](https://github.com/sonic-net/sonic-buildimage/issues/21931)) was `config load_minigraph` / `config reload` — a statistical issue with no special steps required. The race was between bgpcfgd's `vtysh -f /tmp/tmpfile` and concurrent short-lived `vtysh` sessions.

The test correctly uses `config_reload` which matches the production trigger. However, since the bug was statistical (>80% on slow CPUs, rare on fast hardware), consider actively spawning competing `vtysh` sessions during reload to amplify the race window:
```python
# Amplify race: run competing vtysh probes during config_reload
duthost.shell("for i in $(seq 1 20); do docker exec bgp vtysh -c 'show version' & done; wait")
```
Without this, the test is a correctness check (set-src survives reload) rather than an active race inducer. Both have value, but the race amplification would make it a stronger regression guard.

Bojun-Feng replied:
Feedback 3, resolved by https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/370217b0690b0f25f696d85d15186eddf77a5425

Added vtysh race loop spawning competing sessions during `config_reload`.

deepak-singhal0408 commented on tests/bgp/test_frr_set_src_mgmtd.py:
```diff
@@ -0,0 +1,244 @@
+    duthost.shell("sudo sonic-cfggen -j {} --write-to-db".format(BLOAT_CONFIG_TMPFILE))
+    return prefixes
+
+
```
The 200 static routes are loaded by `staticd` from CONFIG_DB, not through bgpcfgd's `vtysh -f` session. They don't extend the `vtysh -f` window where the mgmtd candidate datastore race occurs ([FRRouting/frr#18601](https://github.com/FRRouting/frr/pull/18601)).

If the intent is to widen the race window, consider injecting config directly into the FRR config file (e.g., extra route-maps or prefix-lists) that bgpcfgd loads via `vtysh -f`. That would genuinely extend the vulnerable session duration. See [sonic-mgmt#24061](https://github.com/sonic-net/sonic-mgmt/pull/24061) for an approach that stresses FRR with large config loads.

Bojun-Feng replied:
Feedback 4, resolved by https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/9fe99fb3dd4784ec3da01dd59698485f16b50557

As suggested, we are now injecting 512 FRR prefix-lists + a route-map directly via `vtysh -f` into the bgp container, which should widen the race window.

deepak-singhal0408 commented on tests/bgp/test_frr_set_src_mgmtd.py:
```diff
@@ -0,0 +1,244 @@
+        _verify_route_maps_in_running_config(asichost, lo_ipv4, lo_ipv6)
+
+
+def _find_static_route_anchor(duthost):
```
**Nit**: CodeQL will flag this — some paths return `None` implicitly (fall-through) while others return explicitly. Consider adding an explicit `return None` at the end for clarity.

Bojun-Feng replied:
Feedback 5, resolved by https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/9fe99fb3dd4784ec3da01dd59698485f16b50557

Helper is removed, so issue is nullified.

deepak-singhal0408 commented on tests/bgp/test_frr_set_src_mgmtd.py:
```diff
@@ -0,0 +1,244 @@
+"""Ensure mgmtd FRR replays preserve default-route set-src even with large configs."""
```
**Nit**: Since mgmtd is mandatory in all FRR 10.x (not a toggleable feature), the filename `test_frr_set_src_mgmtd` slightly misleads. Something like `test_frr_set_src_route_map.py` would better describe what's actually being tested — that the `RM_SET_SRC` route-map survives config reload.

Bojun-Feng replied:
Feedback 6, resolved by https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/4aea95d6da30c28b34f536c21920d3c708ceca17

Test file is renamed as suggested. Note first commit referring to feedback 6 (https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/9dac8a0423093a406d4cae6b27c3606d94558f32) renamed the file incorrectly and was patched later.

deepak-singhal0408 commented on tests/bgp/test_frr_set_src_mgmtd.py:
```diff
@@ -0,0 +1,244 @@
+
+def _add_static_routes_for_bloat(duthost, interface, nexthop, count=BLOAT_STATIC_ROUTE_COUNT):
+    """Write a batch of static routes into CONFIG_DB via sonic-cfggen."""
+    prefixes = _generate_bloat_prefixes(count)
```
**Nit**: This per-prefix cleanup loop is slow (200 iterations × shell calls). Since the test uses `config_checkpoint` (line 190) and `config_rollback` (line 237), the rollback already restores the original state. This explicit cleanup may be unnecessary — or could be replaced with a single bulk `sonic-db-cli` pipeline.

Bojun-Feng replied:
Feedback 7, resolved by https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/9fe99fb3dd4784ec3da01dd59698485f16b50557

Since bloat is now FRR prefix-lists (not CONFIG_DB static routes), cleanup is one `no ip prefix-list BLOAT_PL` + `no route-map BLOAT_RM` call per container.

deepak-singhal0408 reviewed (commented):
**Nit / Info-level comments** — not blocking, but worth considering.

**PR Description**: The `Fix sonic-buildimage#24694` reference is stale (that issue was self-filed and already closed Dec 2025). The PR should reference the actual bug [sonic-buildimage#21931](https://github.com/sonic-net/sonic-buildimage/issues/21931) and the upstream fix [FRRouting/frr#18601](https://github.com/FRRouting/frr/pull/18601) / [FRRouting/frr#18541](https://github.com/FRRouting/frr/issues/18541).

Bojun-Feng committed:
[test]: Feedback 1,2 - Fix mgmtd skip condition and add T2 topology

Signed-off-by: Bojun Feng <bojundf@gmail.com>

Bojun-Feng committed:
[test]: Remove redundant wait_critical_processes already called by safe_reload

Signed-off-by: Bojun Feng <bojundf@gmail.com>

Bojun-Feng committed:
[test]: Feedback 3 - Add vtysh race loop for race amplification

Signed-off-by: Bojun Feng <bojundf@gmail.com>

Bojun-Feng committed:
[test]: Feedback 4,5,7 - Replace static-route bloat with FRR prefix-list bloat

* Static routes via CONFIG_DB don't exercise the vtysh -f replay path where
  the mgmtd race occurs; replaced with prefix-lists injected directly into
  FRR via vtysh -f to match the actual vulnerable code path
* Removed _get_remote_ip helper with mixed implicit/explicit returns that
  CodeQL flagged; no longer needed without static-route nexthop lookup
* Replaced 512-iteration sonic-db-cli cleanup loop with single vtysh
  'no route-map' + 'no prefix-list' command

Signed-off-by: Bojun Feng <bojundf@gmail.com>

Bojun-Feng committed:
[test]: Feedback 6 - Rename test file to reflect route-map scope

Signed-off-by: Bojun Feng <bojundf@gmail.com>

Bojun-Feng committed:
[test]: Add wait_until polling for BGP convergence after reload

Signed-off-by: Bojun Feng <bojundf@gmail.com>

Bojun-Feng committed:
[test]: Add protocol binding assertions for RM_SET_SRC route-maps

Signed-off-by: Bojun Feng <bojundf@gmail.com>

Bojun-Feng committed:
[test]: Fix import path and wait for BGP convergence after reload
* Switch from gu_utils to gcu_utils for fail_on_rollback_error kwarg
* Ensure BGP sessions are established before assert route-map presence

Signed-off-by: Bojun Feng <bojundf@gmail.com>

Bojun-Feng committed:
[test]: Use POSIX-safe redirection for race loop background process

Signed-off-by: Bojun Feng <bojundf@gmail.com>

mssonicbld commented:
/azp run

Bojun-Feng committed:
[test]: Use dut_with_default_route fixture for T2 compatibility
* Test verifies set_src on the default route as data-plane proof that the
  route-map is actually applied (not just present in config)
* On T2 chassis, default route is only advertised from T3 to one linecard
* dut_with_default_route (from conftest.py) selects the correct linecard

Signed-off-by: Bojun Feng <bojundf@gmail.com>

mssonicbld commented:
/azp run

Bojun-Feng commented on tests/bgp/test_frr_set_src_mgmtd.py:
```diff
@@ -0,0 +1,244 @@
+        _verify_route_maps_in_running_config(asichost, lo_ipv4, lo_ipv6)
+
+
+def _find_static_route_anchor(duthost):
```
Duplicate of Feedback 5, resolved by https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/9fe99fb3dd4784ec3da01dd59698485f16b50557

Bojun-Feng committed:
[test]: Feedback 6 - Patch incorrectly renamed test file

Signed-off-by: Bojun Feng <bojundf@gmail.com>

mssonicbld commented:
/azp run

Bojun-Feng commented:
Thank you @deepak-singhal0408 for the incredibly thorough review. The depth of analysis taught me a lot. I genuinely appreciate the time and effort that went into this.

Apologies for the delay; I wanted to ensure the changes were solid. Since I significantly rewrote the test, I split the updates into logical commits rather than one large blob. Hope this structure clarifies my reasoning. To make re-review easier, here's how each feedback maps to the commits:

| # | Feedback | Solution | Commit |
|---|----------|----------|--------|
| 1 | `L192`: Skip condition wrong — `frr_mgmt_framework_config` doesn't control mgmtd | Now checks `pgrep -x mgmtd` in bgp containers | https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/8b8a44ed3663529d1c2feccaa6deb1bd8125d5aa |
| 2 | `L16`: T2 topology missing | Added T2 to marker + `dut_with_default_route` fixture | https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/8b8a44ed3663529d1c2feccaa6deb1bd8125d5aa |
| 3 | `L197`: No race amplification | Added vtysh race loop spawning competing sessions during `config_reload` | https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/370217b0690b0f25f696d85d15186eddf77a5425 |
| 4 | `L170`: Static routes don't trigger race | Replaced with FRR prefix-list bloat via `vtysh -f` | https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/9fe99fb3dd4784ec3da01dd59698485f16b50557 |
| 5 | `L108`: Mixed implicit/explicit returns | `_get_remote_ip` removed entirely (CodeQL resolved by deletion) | https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/9fe99fb3dd4784ec3da01dd59698485f16b50557 |
| 6 | `L1`: Filename misleading | Renamed to `test_frr_set_src_route_map.py` | https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/4aea95d6da30c28b34f536c21920d3c708ceca17 |
| 7 | `L157`: Per-prefix cleanup loop slow | Single `no route-map` + `no prefix-list` command | https://github.com/sonic-net/sonic-mgmt/pull/21429/commits/9fe99fb3dd4784ec3da01dd59698485f16b50557 |

Note feedback 1,2 are in one commit; 4,5,7 are also grouped together (same code area); 3 and 6 are standalone.

I have resolved the conversations that are addressed. I've also updated the PR description to reference the original issue and upstream fix.

Let me know if there is anything else that needs to be fixed, thanks!

Bojun-Feng committed:
[test]: Place bloat config in /etc/frr/ instead of /tmp/ for vtysh -f

* FRR 10.x daemons with mgmtd may run with PrivateTmp or separate
  mount namespaces, making /tmp invisible to child processes spawned
  by vtysh -f when distributing config to each daemon (mgmtd, zebra,
  bgpd, staticd)
* Move bloat.conf to /etc/frr/ inside the bgp container which is
  always shared across all FRR daemon processes
* Clean up the temporary file after vtysh applies it

Signed-off-by: Bojun Feng <bojundf@gmail.com>

mssonicbld commented:
/azp run

Bojun-Feng commented:
Hi @deepak-singhal0408 — gentle ping on this. I've addressed all 7 items from your review (commit breakdown in my May 7 comment above) and CI is passing. Would you be able to take another look when you get a chance? Happy to clarify anything further. Thanks!

bhouse-nexthop commented:
/azp run

bhouse-nexthop reviewed (approved):
Approving.

- **Type:** new regression test (mgmtd `set src` after FRR replay).
- **CI runs the test?** Yes — `pytest.mark.topology('t0','t1','t2')`, selected by the default gate. Meaningful assertions run only where mgmtd is live and a default route exists (`pytest_require(_mgmtd_running(...))`), so a real FRR-10.x pass is what fully validates the regression; on KVM the body may skip.
- **Rationale:** all 7 of deepak-singhal0408's review points are addressed at the current head.
- **Issue:** closes #21342 (test-gap) on merge — intended. (The sonic-buildimage#24694 URL reference is cross-repo and won't auto-close.)
- Non-blocking nit: the PR description is stale (mentions the removed `frr_mgmt_framework_config` gating / CONFIG_DB static routes).
